### ISSUE TITLE
Could not connect to the MySQL service. Please check the config file

### ISSUE THREAD

**Issue by dbsheajr** [ethicalhack3r/DVWA#358]
Hello. I have downloaded and installed the DVWA on an Ubuntu server:

***
user@ubuntuserver:~$ uname -a 

Linux ubuntuserver 5.4.0-33-generic #37-Ubuntu SMP Thu May 21 12:53:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux 
***

I read the README.md document and watched the 27 minute YouTube video concerning getting DVWA setup, including changing file & folder permissions, etc. After watching the YT video, all modules (reCAPTCHA key, etc.) are green.

I can get the setup.php file to load at http://<UBUNTU_SERVER_IP>/dvwa/setup.php

When I click in <Create / Reset Database>, I receive the following error: "Could not connect to the MySQL service. Please check the config file."

I can manually authenticate to mysql using the root user (-uroot) as well as an additional administrator account I created (dvwa [-udvwa]).

Partial information from /var/www/html/dvwa/config/config.inc.php:

***
$_DVWA = array(); 
$_DVWA[ 'db_server' ]   = '127.0.0.1'; 
$_DVWA[ 'db_database' ] = 'dvwa'; 
$_DVWA[ 'db_user' ]     = 'dvwa'; 
$_DVWA[ 'db_password' ] = 'p@ssw0rd';
***

I have searched the github > DVWA past issues, but I am not finding the resolution to the issue.

That being said, I did observe something that may be of help. I ran Wireshark and captured HTML traffic. I see that the client is sending "GET /dvwa/setup.php" which is being being replied to with "HTTP/1.1 200 OK", but when the client sends "GET /dvwa/js/add_event_listeners.js", the server responds with "HTTP/1.1 404 Not Found" (see attached WireShark image). I confirmed that this file exists in the path specified and I ran the chmod and chgrp commands on this directory, but that did not resolve the issue. 

I am slightly more knowledgeable about Linux than a totally new user, but not much. I would kindly appreciate any help.

![image](https://user-images.githubusercontent.com/66387738/83685734-21a0b580-a5b7-11ea-9cf7-4bf5897cffbd.png)

**Comment by digininja** [comment]
Get a screenshot connecting to the database on the command line using the
dvwa user with the password from the config file. Use -p and the password
with no space between the two

**Comment by dbsheajr** [comment]
Thank you for the quick reply.

![image](https://user-images.githubusercontent.com/66387738/83686621-83adea80-a5b8-11ea-95ab-c5edc245c89f.png)

**Comment by digininja** [comment]
Is there any reason you are using mysql instead of mariadb?

Mysql tends to cause problems the are harder to diagnose then mariadb.
Unless you specifically need it, I'd swap over.

**Comment by dbsheajr** [comment]
The short answer is "no; there is not a particular reason", but I do have SQL & MySQL experience.

That being said, I don't know what I need to do to switch DVWA from MySQL to mariadb.

**Comment by digininja** [comment]
To swap, nothing. Just uninstall mysql and install mariadb-server.

The other thing to try is change the config file to localhost and the IP
address, there is an issue on redhat distros where the server name matters.

**Comment by dbsheajr** [comment]
Ok, I ran the following commands:

***
sudo apt-get remove mysql
sudo apt-get remove mysqld
sudo apt-get install mariadb-server
***
In the config file, I changed $_DVWA[ 'db_server' ] = 'localhost';

I am still receiving the same error after clicking the create / reset database.

README.md suggests creating a new mariadb user/admin account, but I am unable to connect to mysql as described in the README document.

**Comment by digininja** [comment]
There are no functional differences between mysql and mariadb on the
command line, you can use the two interchangeable.

The root user doesn't have a password by default. Connect at that and then
add the new user as before.

**Comment by dhodyrahmad** [comment]
@dbsheajr did you solved it? Im also using mysql on centos and have same problem with you.

**Comment by digininja** [comment]
Mysql and Centos will be Selinux, search closed tickets, it's been covered
before.

**Comment by dbsheajr** [comment]
@dhodyrahmad, no, I am still working on it.

@digininja: after uninstalling mysql and trying to get mariadb running per your recommendation, I was unable to connect to a DB and I felt I was going backwards. I deleted the Ubuntu server VM and rewatched the YT video and got everything setup again. I'm back to the same problem: Unable to connect to database.

I am still able to manually connect to mysql using the "dvwa" account. I double checked that the settings in the config file match.

I ran tcpdump on the Ubuntu server while filtering for traffic on port 3306. I see bi-directional traffic between the services/daemons. I also see something about "caching sha 2 password". I saw this mentioned in another thread, but I didn't understand how the issue was resolved. Any guidance is appreciated.

![image](https://user-images.githubusercontent.com/66387738/83795504-90404a80-a66d-11ea-8dde-e3fc9838ef5b.png)

**Comment by digininja** [comment]
Mysql uses a different password format when connecting, the standard PHP
module doesn't talk that format do won't connect to it, that's why I
recommend mariadb.

From the command line you shouldn't set anything different to mysql, of you
couldn't connect on command line then something was wrong with either the
install or the passwords you were using. I'd persist with getting that
working at it will be easier in the long run.

Alternatively, the other ticket details how to change the password hashing
format but it's been a while since I looked at it so don't remember any of
it

**Comment by dbsheajr** [comment]
Update:

I did successfully update mysql to mariadb using information from this page:
***
https://mariadb.com/resources/blog/how-to-migrate-from-mysql-to-mariadb-on-linux-in-five-steps/
***
which redirected me to this page:
***
https://downloads.mariadb.org/mariadb/repositories/#distro=Ubuntu&distro_release=focal--ubuntu_focal&mirror=liquidtelecom&version=10.4
(for anyone else reading this looking for assistance, note that the second link is specific to installing  mariadb v. 10.4 [stable] on ubuntu 20.04).
***
I ran the commands listed on the second page (again, according to the page, these are specific to ubuntu 20.04):

sudo apt-get install software-properties-common
sudo apt-key adv --fetch-keys 'https://mariadb.org/mariadb_release_signing_key.asc'
sudo add-apt-repository 'deb [arch=amd64,arm64,ppc64el] http://mariadb.mirror.liquidtelecom.com/repo/10.4/ubuntu focal main'

then: 

sudo apt update
sudo apt install mariadb-server

Finally, I created /etc/apt/sources.list.d/mariadb.list and populated the file with the repositories listed in the document.
***
mysql version now shows to be mariadb:
![image](https://user-images.githubusercontent.com/66387738/83801739-a81ccc00-a677-11ea-9248-c2ba4f1b11a2.png)
***
That being said, after mighrating to mariadb, I can no longer authenticate with mysql:
![image](https://user-images.githubusercontent.com/66387738/83802032-13669e00-a678-11ea-86a1-7a6da74f1fb3.png)
***
I ran into this issue earlier before I migrated to mariadb and the fix was to run "sudo apt-get install mysql-server". It seems like I am going in circles.

@digininja: in your previous reply, there are some typos that made your reply a little hard to follow along with. To clarify, before migrating from mysql to mariadb, I was able to authenticate using "mysql -udvwa -pp@ssw0rd". After migrating to mariadb, I am unable to do so

I tried starting and restarting the mariadb service to no avail:

![image](https://user-images.githubusercontent.com/66387738/83802920-99cfaf80-a679-11ea-9efa-60b594615314.png)

Here is the command it suggests I run:

![image](https://user-images.githubusercontent.com/66387738/83803045-d7ccd380-a679-11ea-998f-9d4d4f1243c5.png)

I have spent about 10 hours trying to get the database to communicate. Any specific information that will help resolve this issue is greatly appreciated.

**Comment by digininja** [comment]
Don't start installing from custom sources, use the Ubuntu package for
mariadb, it works out of the box. You seem to be over complicating this
which isn't going to help get things working.

The user that you created on the mysql database will be gone as it is part
of the database. When you installed mariadb you start again with a blank
permissions table so root and a blank password should work. There is also a
script to configure the database. I can't remember the name but it talks
you through locking it down and adding passwords.

Something to watch out for is by default, you can only log in as root
database user while you are the root Linux user. That is why we have to
tell people to add their own user as the web app won't be running as root
so can't talk to the database as root

**Comment by dbsheajr** [comment]
Do you think you can help me uncomplicate this?

Let's say you watched the YT video referenced in the README.md file. After watching the video, you have the DVWA files on a VM Ubuntu Server that you are accessing from a Kali VM. The setup.php pages displays but will not connect to the database.

Also, you can connect to the MySQL database manually using "mysql -udvwa -pp@ssw0rd"
***
With the above scenario described, your recommended to replace mysql with mariadb.
***
If you were in my shoes, what **exactly** (step-by-step) would you do go covert mysql to mariadb?

I have googled, I have watched YT videos, I have read other tickets in the DVWA GitHub repository. After two days of googling, watching YT videos, and reading tickets **I am no further ahead than I was on Wednesday***. Are you able to provide specific help? If not, is there someone else that can provide specific help? Based off of all the tickets created with similar subject lines as mine, I have a feeling this is probably a quick fix.

**Comment by digininja** [comment]
I'd start with a new VM if I were you.

If not, remove all the junk you added to add mariadb, all the extra apt
sources etc.

Then remove the database servers

apt remove mariadb* mysql*

apt autoremove

update the sources

apt update

install the standard mariadb packages from the distro

apt install mariadb-server mariadb-client

mysql -u root

create the user as covered in the readme and you are good to go. Don't
worry about the warnings on the setup page, they are for specific areas
that aren't used that often.

**Comment by dbsheajr** [comment]
That is what I needed! It appears to be working now. Thank you so much. Closing ticket.
